File: 150-EvaluateReversePolishNotation/150-EvaluateReversePolishNotation.py
# Last updated: 07/04/2026, 11:40:50
import math

class Solution:
    def evalRPN(self, tokens: List[str]) -> int:
        # Initialize an empty stack to keep track of operands
        stack = []
        
        # Define the set of operators
        operators = {'+', '-', '*', '/'}
        
        # If there's only one token, return its integer value
        if len(tokens) == 1: 
            return int(tokens[0])
        
        # Iterate through each token in the input list
        for token in tokens:
            # If the token is not an operator, push it as an integer onto the stack
            if token not in operators:
                stack.append(int(token))
            else:
                # If it's an operator, pop the top two operands from the stack
                sec_operand = stack.pop()
                fir_operand = stack.pop()
                tmp = 0
                
                # Perform the corresponding operation based on the operator
                if token == '+':
                    tmp = fir_operand + sec_operand
                elif token == '-':
                    tmp = fir_operand - sec_operand
                elif token == '*':
                    tmp = fir_operand * sec_operand
                elif token == '/':
                    # Perform integer division and truncate the result
                    tmp = fir_operand / sec_operand
                    tmp = math.trunc(tmp)
                
                # Push the result back onto the stack
                stack.append(tmp)
        
        # The final result is the only element left on the stack
        return stack.pop()

File: 150-EvaluateReversePolishNotation/150-EvaluateReversePolishNotation.java
// Last updated: 01/05/2026, 14:03:12
1class Solution {
2
3    private Stack<String> store;
4    private boolean isOperand (String s){
5
6        if ( s.equals("+") || s.equals("/") ||
7        s.equals("*") ||s.equals("-") ) return true;
8        return false;
9    }
10    private String operation(String a , String b, String s){
11        
12        if (  s.equals("+")) return Integer.toString(
13            Integer.parseInt(a)+Integer.parseInt(b)
14        );
15         else if (  s.equals("-")) return Integer.toString(
16            Integer.parseInt(a)-Integer.parseInt(b)
17        );
18         else if (  s.equals("/")) return Integer.toString(
19            Integer.parseInt(a)/Integer.parseInt(b)
20        );
21        else{  return Integer.toString(
22            Integer.parseInt(a)*Integer.parseInt(b)
23        );}
24
25        
26        
27    }
28
29    
30    public int evalRPN(String[] tokens) {
31
32        store = new Stack<>();
33
34        for (String token : tokens){
35            if ( isOperand(token)){
36                String a = store.pop();
37                String b = store.pop();
38                // b goes first into stack;
39                String res = operation(b,a,token);
40                store.push(res);
41            }else{
42                store.push(token);
43            }
44        }
45
46        return Integer.parseInt(store.peek());
47        
48        
49
50    }
51}